src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,890 @@
+    // my first instruction via a fetch request.
+
+    _committedState->activate();
+
+    auto callback = [this]{
+        advanceInst();
+    };
+
+    // Schedule the first call to the entry point of thread runtime. We ensure
+    // a clean callstack by sending the call through the scheduler as a
+    // separate event.
+    _cpuPtr->schedule(new EventFunctionWrapper(callback,
+                                               name() + ".activateEvent",
+                                               true),
+                      curTick());
+}
+
+void
+SDCPUThread::advanceInst()
+{
+    DPRINTF(SDCPUThreadEvent, "advanceInst()\n");
+
+    const InstSeqNum seq_num = lastCommittedInstNum + inflightInsts.size() + 1;
+
+    TheISA::PCState next_pc = getNextPC();
+
+    if (isRomMicroPC(next_pc.microPC())) {
+        const StaticInstPtr static_inst =
+            _cpuPtr->microcodeRom.fetchMicroop(next_pc.microPC(), curMacroOp);
+
+        DPRINTF(SDCPUInstEvent,
+                "Decoded ROM microop (seq %d) - %#x : %s\n",
+                seq_num,
+                next_pc.microPC(),
+                static_inst->disassemble(next_pc.microPC()).c_str());
+
+        shared_ptr<InflightInst> dynamic_inst_ptr =
+            make_shared<InflightInst>(getThreadContext(), &memIface, seq_num,
+                                      next_pc, static_inst);
+
+        inflightInsts.push_back(dynamic_inst_ptr);
+
+        issueInstruction(dynamic_inst_ptr);
+
+        return;
+    }
+
+    if (curMacroOp) {
+        const StaticInstPtr static_inst =
+            curMacroOp->fetchMicroop(next_pc.microPC());
+
+        DPRINTF(SDCPUInstEvent,
+                "Decoded microop (seq %d) - %#x : %s\n",
+                seq_num,
+                next_pc.microPC(),
+                static_inst->disassemble(next_pc.microPC()).c_str());
+
+        shared_ptr<InflightInst> dynamic_inst_ptr =
+            make_shared<InflightInst>(getThreadContext(), &memIface, seq_num,
+                                      next_pc, static_inst);
+
+        inflightInsts.push_back(dynamic_inst_ptr);
+
+        issueInstruction(dynamic_inst_ptr);
+
+        return;
+    }
+
+    shared_ptr<InflightInst> dynamic_inst_ptr =
+        make_shared<InflightInst>(getThreadContext(), &memIface, seq_num,
+                                  next_pc);
+
+    inflightInsts.push_back(dynamic_inst_ptr);
+
+    fetchOffset = 0;
+    onBeginFetch(weak_ptr<InflightInst>(dynamic_inst_ptr));
+}
+
+void
+SDCPUThread::commitAllAvailableInstructions()
+{
+    shared_ptr<InflightInst> inst_ptr;
+
+    while (!inflightInsts.empty()
+        && ((inst_ptr = inflightInsts.front())->isComplete()
+            || inst_ptr->isSquashed())) {
+
+        if (inst_ptr->fault() != NoFault) {
+            handleFault(inst_ptr);
+
+            return;
+        }
+
+        if (!inst_ptr->isSquashed()){
+            commitInstruction(inst_ptr);
+        }
+
+        inflightInsts.pop_front();
```

Comment:
For a squashed instruction, is there any cleanup you need to worry about? I guess not since you're passing around weak pointers everywhere and the instruction will be deleted. Could you make a note of that here?